presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/LimitPushDown.java

```diff
@@ -134,13 +142,40 @@ public PlanNode visitAggregation(AggregationNode node, RewriteContext<LimitConte
         {
             LimitContext limit = context.get();
 
-            if (limit != null &&
-                    node.getAggregations().isEmpty() &&
+            if (limit != null) {
+                checkArgument(!node.getSampleWeight().isPresent(), "DISTINCT aggregation has sample weight symbol");
+                if (node.getAggregations().isEmpty() &&
                     node.getOutputSymbols().size() == node.getGroupBy().size() &&
                     node.getOutputSymbols().containsAll(node.getGroupBy())) {
-                checkArgument(!node.getSampleWeight().isPresent(), "DISTINCT aggregation has sample weight symbol");
-                PlanNode rewrittenSource = context.rewrite(node.getSource());
-                return new DistinctLimitNode(idAllocator.getNextId(), rewrittenSource, limit.getCount(), false, Optional.empty());
+                    PlanNode rewrittenSource = context.rewrite(node.getSource());
+                    return new DistinctLimitNode(idAllocator.getNextId(), rewrittenSource, limit.getCount(), false, Optional.empty(), node.getGroupBy());
+                }
+                else if (node.getFunctions().values().stream().allMatch(signature -> signature.getName().equals("arbitrary"))) {
+                    PlanNode rewrittenSource = context.rewrite(node.getSource());
+                    ImmutableMap.Builder<Symbol, Expression> aggregationMapBuilder = ImmutableMap.builder();
+                    for (Entry<Symbol, FunctionCall> entry : node.getAggregations().entrySet()) {
+                        aggregationMapBuilder.put(entry.getKey(), entry.getValue().getArguments().get(0));
+                    }
+                    Map<Symbol, Expression> aggregationSymbolMap = aggregationMapBuilder.build();
+
+                    List<Symbol> nonDistinctSymbols = aggregationSymbolMap.values().stream()
+                        .map(expression -> Symbol.fromQualifiedName(((QualifiedNameReference) expression).getName()))
+                        .collect(toList());
+
+                    PlanNode filterNode = rewrittenSource;
+                    for (Symbol nonDistinctSymbol : nonDistinctSymbols) {
+                        filterNode = new FilterNode(idAllocator.getNextId(), rewrittenSource, new IsNotNullPredicate(nonDistinctSymbol.toQualifiedNameReference()));
```

Comment:
Hmm, I don't think this is correct. This has the potential to filter out all of the rows when the original aggregation would not have (e.g. if the arbitrary() columns are never all not null at the same time).